File: Challenge/Math/gcd/gcd.cpp
#include <iostream>

using namespace std;

int gcd(int a, int b) {
    return b ? gcd(b, a%b) : a;
}

int extgcd(int a, int b, int &x, int &y) {
    int d = a;
    
    if (b) {
        d = extgcd(b, a%b , y, x);
        // y -= (a/b)*x;
        // y = y - (a/b)*x;
        y = x-(a/b)*b*y;
    } else {
        x = 1;
        y = 0;
    }
    return d;
}

int main() {
    int x, y;
    x = y = -10086;
    cout << gcd(4,11) << endl;
    cout << extgcd(4, 11, x, y) << endl;
    cout << x << y << endl;
    return 0;
}

File: Challenge/Math/gcd/gcd.py
# 最大公因数
def gcd(a, b):
    return a if b == 0 else gcd(b, a%b)

print(gcd(12,15))
print(gcd(12,0))

# 扩张欧几里得算法
def extgcd(a, b, x, y):
    d = a
    if b != 0:
        d = extgcd(b, a%b,y,x)
        y -= (a/b)*x
    else:
        x = 1
        y = 0
    return d

x = y = 0
extgcd(4,11,x,y)
